starlark/hashtable.go

```diff
@@ -371,3 +364,186 @@ func softHashString(s string) uint32 {
 	}
 	return h
 }
+
+// An OrderedStringDict is a mapping from names to values with support
```

Comment:
This code is all very well, but I wonder why it needs to be in this package, rather than in your application? It doesn't use any internals of the starlark package (other than `overloaded`, which is tiny, and hashString, which is exported as String.Hash).

If this exists only for the implementation of Struct, then let's move it into that package, but mutability and insertion ordering don't seem relevant to struct.

If a general need for a mutable insertion-ordered struct-like type arises across several projects, you could contribute your implementation to the lib/ tree, but it seems premature to put it here to start with.

Reply:
My thinking was that most applications using starlark will implement custom objects they need to pass in and out of starlark scripts. My goto has been `starlark.StringDict` but then it does make it a little difficult to implement comparisons other methods. Seems like the missing type is hashmap with iteration support.

I modified starlarkstruct to showcase how it would bind to the new type.

I agree this is probably premature to add now, will use in my application to see more use cases, happy to close!